query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java

```diff
@@ -189,6 +369,58 @@ public void handleRequest(
     }
   }
 
+  @Nonnull
+  private QueryRequest rewriteRequestWithViewFiltersApplied(QueryRequest request,
+      Map<String, ViewColumnFilter> columnFilterMap) {
+
+    Filter newFilter = removeViewColumnFilter(request.getFilter(), columnFilterMap);
+    QueryRequest.Builder builder = QueryRequest.newBuilder(request).clearFilter();
+    if (!Filter.getDefaultInstance().equals(newFilter)) {
+      builder.setFilter(newFilter);
+    }
+    return builder.build();
+  }
+
+  private Filter removeViewColumnFilter(Filter filter,
+      Map<String, ViewColumnFilter> columnFilterMap) {
+    if (filter.getChildFilterCount() > 0) {
+      // Recursively try to remove the filter and eliminate the null nodes.
+      Set<Filter> newFilters = filter.getChildFilterList().stream()
+          .map(f -> removeViewColumnFilter(f, columnFilterMap))
+          .filter(f -> !Filter.getDefaultInstance().equals(f))
+          .collect(Collectors.toCollection(LinkedHashSet::new));
+
+      if (newFilters.isEmpty()) {
+        return Filter.getDefaultInstance();
+      } else if (newFilters.size() == 1) {
+        return newFilters.stream().findFirst().get();
```

Comment:
nit - believe we're already using guava in here, so a little clearer to do:
```suggestion
        return Iterables.getOnlyElement(arguments);
```

Reply:
Oh this is cool. Will fix it.